// @mdp-framework/adapters/dist/interfaces.js
/**
 * Adapter Interfaces (FROZEN as of M1.5)
 *
 * These interfaces define the contracts between Team A (Registry/Resolver)
 * and Team B (Action Engine) for provider adapters.
 *
 * BREAKING CHANGES require:
 * 1. Major version bump
 * 2. Cross-team coordination
 * 3. Migration guide in CONTRACTS.md
 *
 * @since M1.5 (Contract Hardening)
 */
// ============================================================================
// ERROR TAXONOMY (FROZEN)
// ============================================================================
/**
 * Standard provider error codes
 *
 * These codes allow Team B (Action Engine) to handle failures uniformly
 * without knowing provider-specific error details.
 */
export var ProviderErrorCode;
(function (ProviderErrorCode) {
    // Authentication/Authorization (40x)
    ProviderErrorCode["AUTH_MISSING"] = "AUTH_MISSING";
    ProviderErrorCode["AUTH_INVALID"] = "AUTH_INVALID";
    ProviderErrorCode["AUTH_EXPIRED"] = "AUTH_EXPIRED";
    ProviderErrorCode["PERMISSION_DENIED"] = "PERMISSION_DENIED";
    // Resource errors (404, 410)
    ProviderErrorCode["NOT_FOUND"] = "NOT_FOUND";
    ProviderErrorCode["GONE"] = "GONE";
    // Rate limiting (429)
    ProviderErrorCode["RATE_LIMIT"] = "RATE_LIMIT";
    // Server errors (50x)
    ProviderErrorCode["SERVER_ERROR"] = "SERVER_ERROR";
    ProviderErrorCode["SERVICE_UNAVAILABLE"] = "SERVICE_UNAVAILABLE";
    // Network/connectivity
    ProviderErrorCode["NETWORK_ERROR"] = "NETWORK_ERROR";
    ProviderErrorCode["TIMEOUT"] = "TIMEOUT";
    // Validation
    ProviderErrorCode["INVALID_INPUT"] = "INVALID_INPUT";
    ProviderErrorCode["SCHEMA_VALIDATION_FAILED"] = "SCHEMA_VALIDATION_FAILED";
    // Unknown/Other
    ProviderErrorCode["UNKNOWN"] = "UNKNOWN";
})(ProviderErrorCode || (ProviderErrorCode = {}));
/**
 * Standardized provider error
 *
 * Adapters MUST map provider-specific errors to this taxonomy.
 */
export class ProviderError extends Error {
    code;
    details;
    constructor(code, message, details) {
        super(message);
        this.code = code;
        this.details = details;
        this.name = 'ProviderError';
        Error.captureStackTrace?.(this, ProviderError);
    }
    /**
     * Whether this error is retryable
     */
    get isRetryable() {
        return this.details?.retryable ?? this.isRetryableByCode();
    }
    isRetryableByCode() {
        const retryableCodes = [
            ProviderErrorCode.RATE_LIMIT,
            ProviderErrorCode.SERVER_ERROR,
            ProviderErrorCode.SERVICE_UNAVAILABLE,
            ProviderErrorCode.NETWORK_ERROR,
            ProviderErrorCode.TIMEOUT,
        ];
        return retryableCodes.includes(this.code);
    }
}
