// @mdp-framework/gateway/dist/aoi/index.js
/**
 * AOI (Architecture of Intent) Coordinator
 *
 * Stream 4 implementation - Complete AOI workflow:
 * 1. Vector similarity search and component detection
 * 2. AOI spec generation and owner routing
 * 3. Counter-PR creation
 * 4. Follow-up PR and integration tests
 */
// Core coordinator
export { AOICoordinator } from './coordinator.js';
// Similarity search
export { VectorSearch } from './similarity/vector-search.js';
export { ComponentIndex } from './similarity/component-index.js';
export { ThresholdChecker } from './similarity/threshold-checker.js';
// Spec generation
export { SpecGenerator } from './spec-generator.js';
// Routing
export { OwnerResolver } from './routing/owner-resolver.js';
export { NotificationDispatcher } from './routing/notification-dispatcher.js';
// Counter-PR
export { CounterPRGenerator } from './counter-pr/pr-generator.js';
export { GitHubClient } from './counter-pr/github-client.js';
// Follow-up PR
export { MigrationGenerator } from './followup-pr/migration-generator.js';
export { FollowupPRCreator } from './followup-pr/pr-creator.js';
