// @mdp-framework/gateway/dist/aoi/routing/notification-dispatcher.js
/**
 * Notification Dispatcher
 *
 * Dispatches AOI notifications to Jira, Slack, and other platforms.
 */
export class NotificationDispatcher {
    config;
    constructor(config) {
        this.config = config;
    }
    /**
     * Dispatch notifications for an AOI spec
     */
    async dispatch(spec, policy) {
        const actions = [];
        // 1. Create Jira issue
        if (this.config?.jiraEnabled && this.config.jiraClient) {
            try {
                const issueAction = await this.createJiraIssue(spec, policy);
                actions.push(issueAction);
            }
            catch (error) {
                console.error('Failed to create Jira issue:', error);
                actions.push(this.createFailedAction('create_issue', 'jira', error));
            }
        }
        else {
            // Simulated Jira issue creation
            actions.push(this.createSimulatedIssueAction(spec));
        }
        // 2. Send Slack notification
        if (this.config?.slackEnabled && this.config.slackClient) {
            try {
                const notifyAction = await this.sendSlackNotification(spec, policy);
                actions.push(notifyAction);
            }
            catch (error) {
                console.error('Failed to send Slack notification:', error);
                actions.push(this.createFailedAction('notify', 'slack', error));
            }
        }
        else {
            // Simulated Slack notification
            actions.push(this.createSimulatedNotifyAction(spec));
        }
        return actions;
    }
    /**
     * Create Jira issue (implementation placeholder)
     */
    async createJiraIssue(spec, policy) {
        // In a real implementation, this would call Jira API
        const issueKey = `AOI-${Math.floor(Math.random() * 10000)}`;
        return {
            type: 'create_issue',
            status: 'completed',
            target: {
                platform: 'jira',
                resource_id: issueKey,
                url: `https://jira.example.com/browse/${issueKey}`,
            },
            payload: { spec, policy },
            metadata: {
                created_at: new Date().toISOString(),
                completed_at: new Date().toISOString(),
            },
        };
    }
    /**
     * Send Slack notification (implementation placeholder)
     */
    async sendSlackNotification(spec, policy) {
        // In a real implementation, this would call Slack API
        const threadId = `T${Math.floor(Math.random() * 1000000)}`;
        return {
            type: 'notify',
            status: 'completed',
            target: {
                platform: 'slack',
                resource_id: threadId,
                url: `https://slack.example.com/archives/${threadId}`,
            },
            payload: {
                spec,
                policy,
                message: this.buildSlackMessage(spec),
            },
            metadata: {
                created_at: new Date().toISOString(),
                completed_at: new Date().toISOString(),
            },
        };
    }
    /**
     * Build Slack message for AOI notification
     */
    buildSlackMessage(spec) {
        const owners = spec.owners.join(', ');
        return `
🚀 *New AOI Specification Generated*

*Component:* ${spec.name}
*Description:* ${spec.description}
*Framework:* ${spec.framework}
*Owners:* ${owners}

*Trigger:* ${spec.metadata.trigger_pr}
*Violation:* ${spec.metadata.trigger_violation.message}

Please review the specification and approve for implementation.
    `.trim();
    }
    /**
     * Create simulated issue action (for testing)
     */
    createSimulatedIssueAction(spec) {
        return {
            type: 'create_issue',
            status: 'completed',
            target: {
                platform: 'jira',
                resource_id: `AOI-SIM-${Date.now()}`,
                url: `https://jira.example.com/browse/AOI-SIM-${Date.now()}`,
            },
            payload: { spec },
            metadata: {
                created_at: new Date().toISOString(),
                completed_at: new Date().toISOString(),
            },
        };
    }
    /**
     * Create simulated notify action (for testing)
     */
    createSimulatedNotifyAction(spec) {
        return {
            type: 'notify',
            status: 'completed',
            target: {
                platform: 'slack',
                resource_id: `thread-${Date.now()}`,
                url: `https://slack.example.com/archives/C123456/p${Date.now()}`,
            },
            payload: {
                spec,
                message: this.buildSlackMessage(spec),
            },
            metadata: {
                created_at: new Date().toISOString(),
                completed_at: new Date().toISOString(),
            },
        };
    }
    /**
     * Create failed action
     */
    createFailedAction(type, platform, error) {
        return {
            type,
            status: 'failed',
            target: {
                platform,
            },
            payload: {},
            metadata: {
                created_at: new Date().toISOString(),
                error: error instanceof Error ? error.message : String(error),
            },
        };
    }
}
