// @mdp-framework/mlang/dist/commands/init.js
import { mkdirSync, writeFileSync, existsSync } from 'fs';
import { resolve } from 'path';
export async function runInit(logger, options = {}) {
    const { force = false, verbose = false } = options;
    logger.info('Initializing MLang project...\n');
    const cwd = process.cwd();
    const mlangDir = resolve(cwd, '.mlang');
    // Check if .mlang already exists
    if (existsSync(mlangDir) && !force) {
        logger.error('.mlang directory already exists. Use --force to reinitialize.');
        throw new Error('.mlang directory already exists');
    }
    // Create directory structure
    const dirs = [
        'policies', // NEW: User-friendly policy source folder
        '.mlang',
        '.mlang/keys',
        '.mlang/packs',
        '.mlang/packs/stable',
        '.mlang/packs/preview',
        '.mlang/packs/dev',
        '.mlang/store',
        '.mlang/cache',
        '.mlang/agents',
        '.mlang/agents/ci',
        '.mlang/agents/ci/templates',
    ];
    for (const dir of dirs) {
        const fullPath = resolve(cwd, dir);
        if (!existsSync(fullPath)) {
            mkdirSync(fullPath, { recursive: true });
            if (verbose) {
                logger.debug(`Created directory: ${dir}`);
            }
        }
    }
    logger.success('Created .mlang directory structure');
    logger.success('Created policies/ directory');
    // Create default config.yaml
    const configPath = resolve(mlangDir, 'config.yaml');
    const configContent = `# MLang Configuration
version: v1

# Policy resolution
resolver:
  # Channel to use for policy resolution (stable, preview, dev)
  channel: stable

  # Registry provider (file-based by default)
  registry:
    type: file
    path: .mlang/packs

  # Last Known Good (LKG) fallback
  lkg:
    enabled: true
    ttl: 86400  # 24 hours in seconds

# State storage
state:
  # DSN for state capsule storage
  # file:// for local filesystem, postgresql:// for Postgres
  dsn: file://.mlang/store

  # Retention policy
  retention:
    default_ttl: 2592000  # 30 days
    prune_interval: 86400  # daily

# Policy signing
signing:
  # Require signed policies
  require_signatures: false

  # Public keys directory for verification
  keys_dir: .mlang/keys

# Observability
observability:
  # Enable verbose mode
  verbose: false

  # Enable trace mode
  trace: false

  # Event sink (stdout, jsonl, otel)
  sink: stdout

  # Redact sensitive data
  redact: true

# Execution mode
execution:
  # Mode: enforce, report, dry-run
  mode: report

  # Allow enforcement (gating policies)
  allow_enforcement: false

# Role-to-policy mappings
roles:
  pr:
    policies:
      - eng.pr.review
      - eng.pr.description

  feature:
    policies:
      - eng.feature.discovery

  issue:
    policies:
      - eng.issue.debug
`;
    writeFileSync(configPath, configContent, 'utf-8');
    logger.success('Created config.yaml');
    // Create .gitignore for .mlang
    const gitignorePath = resolve(mlangDir, '.gitignore');
    const gitignoreContent = `# MLang runtime directories
cache/
store/
packs/

# Environment variables (contains secrets!)
.env

# Private signing keys
keys/*.key
keys/*.pem
*.private

# Temporary files
*.tmp
*.log
`;
    writeFileSync(gitignorePath, gitignoreContent, 'utf-8');
    logger.success('Created .mlang/.gitignore');
    // Create .env template
    const envPath = resolve(mlangDir, '.env');
    const envContent = `# MLang Environment Variables
# This file contains sensitive configuration for MLang CLI
# DO NOT commit this file to version control!

# Anthropic API Key (required for AI-powered policy evaluation)
# Get your key from: https://console.anthropic.com/
ANTHROPIC_API_KEY=

# Optional: GitHub Token (required for PR operations)
# GITHUB_TOKEN=

# Optional: AI Configuration
# MLANG_AI_CONFIDENCE_THRESHOLD=0.7
# MLANG_AI_TEMPERATURE=0.1
`;
    writeFileSync(envPath, envContent, 'utf-8');
    logger.success('Created .mlang/.env template');
    // Create README
    const readmePath = resolve(mlangDir, 'README.md');
    const readmeContent = `# MLang Configuration

This directory contains MLang framework configuration and runtime data.

## Structure

- \`config.yaml\` - Main configuration file
- \`keys/\` - Signing keys for policy verification (commit public keys only)
- \`packs/\` - Compiled policy registry (build output)
  - \`stable/\` - Stable channel policies
  - \`preview/\` - Preview channel policies
  - \`dev/\` - Development channel policies
- \`store/\` - State capsule storage (file-based)
- \`cache/\` - Resolver cache (LKG)
- \`agents/\` - Agent pattern templates

## Next Steps

1. Write your policies in the \`policies/\` folder (Markdown format)
2. Build them: \`mlang policy build\`
3. Generate signing keys: \`mlang keys generate\`
4. Run doctor check: \`mlang doctor\`

For more information, see: https://docs.mdp.dev
`;
    writeFileSync(readmePath, readmeContent, 'utf-8');
    logger.success('Created .mlang/README.md');
    // Create placeholder for PR template
    const prTemplatePath = resolve(mlangDir, 'agents/ci/templates/pr-comment.md');
    const prTemplateContent = `# MDP Governance Report

**Status:** {{status}}
**Mode:** {{mode}}
**Duration:** {{duration_ms}}ms

## Policy Results

{{#each results}}
### {{policy_id}}
- **Outcome:** {{outcome}}
{{#if violations}}
- **Violations:**
{{#each violations}}
  - {{message}} ({{location.file}}:{{location.line}})
{{/each}}
{{/if}}
{{/each}}

{{#if actions}}
## Suggested Actions
{{#each actions}}
- [{{type}}] {{message}}
{{/each}}
{{/if}}

---
*Generated by MLang v{{version}}*
`;
    writeFileSync(prTemplatePath, prTemplateContent, 'utf-8');
    logger.success('Created PR comment template');
    logger.info('\n' + '='.repeat(60));
    logger.success('MLang initialized successfully!');
    logger.info('\nNext steps:');
    logger.info('  1. Add your ANTHROPIC_API_KEY to: .mlang/.env');
    logger.info('  2. Write policies in: policies/');
    logger.info('  3. Build them: mlang policy build');
    logger.info('  4. Generate keys: mlang keys generate');
    logger.info('  5. Run doctor: mlang doctor\n');
    logger.info('💡 Tip: Get your API key from https://console.anthropic.com/');
    logger.info('💡 Tip: .mlang/.env is auto-loaded and git-ignored\n');
}
