// @mdp-framework/mcp-server/dist/server.js
/**
 * MCP Server - Main server implementation
 */
import { MCPProtocolHandler } from './protocol/handler.js';
import { StdioTransport, HTTPTransport } from './protocol/transport.js';
import { PolicyGetTool, PolicySearchTool, MemoryGetTool, MemorySearchTool, EvidenceRecordTool, } from './tools/index.js';
/**
 * MCPServer - Main server class
 *
 * Implements the MCP protocol and exposes MDP capabilities as tools
 */
export class MCPServer {
    handler;
    transport;
    startTime;
    currentMode;
    currentPort;
    resolver;
    store;
    constructor(config) {
        this.resolver = config.resolver;
        this.store = config.store;
        this.handler = new MCPProtocolHandler();
        // Auto-register default tools
        if (config.autoRegisterTools !== false) {
            this.registerDefaultTools();
        }
    }
    /**
     * Register a tool handler
     */
    registerTool(handler) {
        this.handler.registerTool(handler);
    }
    /**
     * Unregister a tool handler
     */
    unregisterTool(name) {
        return this.handler.unregisterTool(name);
    }
    /**
     * Get all registered tools
     */
    getTools() {
        return this.handler.getTools();
    }
    /**
     * Start the server
     *
     * @param mode - Transport mode ('stdio' or 'http')
     * @param port - Port for HTTP mode (optional)
     */
    async start(mode, port) {
        if (this.transport?.isRunning()) {
            throw new Error('Server is already running');
        }
        // Create transport
        if (mode === 'stdio') {
            this.transport = new StdioTransport((req) => this.handler.handleRequest(req));
        }
        else if (mode === 'http') {
            if (!port) {
                throw new Error('Port is required for HTTP mode');
            }
            this.transport = new HTTPTransport(port, (req) => this.handler.handleRequest(req));
            this.currentPort = port;
        }
        else {
            throw new Error(`Unsupported transport mode: ${mode}`);
        }
        this.currentMode = mode;
        this.startTime = Date.now();
        await this.transport.start();
    }
    /**
     * Stop the server
     */
    async stop() {
        if (this.transport) {
            await this.transport.stop();
            this.transport = undefined;
            this.currentMode = undefined;
            this.currentPort = undefined;
            this.startTime = undefined;
        }
    }
    /**
     * Get server status
     */
    getStatus() {
        return {
            running: this.transport?.isRunning() || false,
            mode: this.currentMode || 'stdio',
            port: this.currentPort,
            toolCount: this.handler.getTools().length,
            uptime: this.startTime ? Date.now() - this.startTime : 0,
        };
    }
    /**
     * Register default MDP tools
     */
    registerDefaultTools() {
        // Policy tools
        this.registerTool(new PolicyGetTool(this.resolver));
        this.registerTool(new PolicySearchTool(this.resolver));
        // Memory tools
        this.registerTool(new MemoryGetTool(this.store));
        this.registerTool(new MemorySearchTool(this.store));
        // Evidence tools
        this.registerTool(new EvidenceRecordTool(this.store));
    }
}
/**
 * Create an MCP server instance
 *
 * @param config - Server configuration
 * @returns MCPServer instance
 */
export function createMCPServer(config) {
    return new MCPServer(config);
}
