// @mdp-framework/gateway/dist/aoi/counter-pr/pr-generator.js
/**
 * Counter-PR Generator
 *
 * Generates code and creates pull requests in framework repositories.
 */
export class CounterPRGenerator {
    aiProvider;
    constructor(aiProvider) {
        this.aiProvider = aiProvider;
    }
    /**
     * Generate component implementation code
     */
    async generateImplementation(spec, includeTests) {
        const files = [];
        // 1. Generate main component file
        const componentCode = await this.generateComponentCode(spec);
        files.push({
            path: this.inferComponentPath(spec),
            content: componentCode,
            operation: 'create',
        });
        // 2. Generate tests if requested
        if (includeTests) {
            const testCode = await this.generateTestCode(spec);
            files.push({
                path: this.inferTestPath(spec),
                content: testCode,
                operation: 'create',
            });
        }
        return files;
    }
    /**
     * Generate component code using LLM
     */
    async generateComponentCode(spec) {
        const prompt = this.buildImplementationPrompt(spec);
        const response = await this.aiProvider.complete({
            prompt,
            temperature: 0.3, // Lower temperature for code generation
            max_tokens: 3000,
        });
        // Extract code from response (remove markdown fences if present)
        return this.extractCode(response.content);
    }
    /**
     * Generate test code using LLM
     */
    async generateTestCode(spec) {
        const prompt = this.buildTestPrompt(spec);
        const response = await this.aiProvider.complete({
            prompt,
            temperature: 0.3,
            max_tokens: 2000,
        });
        return this.extractCode(response.content);
    }
    /**
     * Build implementation prompt
     */
    buildImplementationPrompt(spec) {
        const examplesSection = spec.examples
            .map((ex) => `
// ${ex.context || 'Example'}
// File: ${ex.file}
\`\`\`${ex.language}
${ex.code}
\`\`\`
`)
            .join('\n');
        return `You are implementing a component for a framework based on the following specification.

Component Name: ${spec.name}
Description: ${spec.description}

Specification:
${spec.specification}

Examples:
${examplesSection}

Task:
Generate a complete, production-ready implementation of this component.

Requirements:
1. Follow the framework's coding conventions
2. Include comprehensive documentation comments
3. Implement all public APIs described in the specification
4. Add proper error handling
5. Make the code maintainable and well-structured

Generate ONLY the code, no explanations. Use the appropriate language based on the framework.`;
    }
    /**
     * Build test prompt
     */
    buildTestPrompt(spec) {
        return `You are writing tests for a component based on the following specification.

Component Name: ${spec.name}
Description: ${spec.description}

Specification:
${spec.specification}

Task:
Generate comprehensive unit tests for this component.

Requirements:
1. Test all public APIs
2. Include edge cases and error conditions
3. Follow the framework's testing conventions
4. Use appropriate test framework (XCTest, Jest, etc.)
5. Make tests readable and maintainable

Generate ONLY the test code, no explanations.`;
    }
    /**
     * Extract code from LLM response (remove markdown fences)
     */
    extractCode(content) {
        // Remove markdown code fences if present
        const codeBlockRegex = /```[\w]*\n([\s\S]*?)\n```/;
        const match = content.match(codeBlockRegex);
        if (match) {
            return match[1].trim();
        }
        return content.trim();
    }
    /**
     * Infer component file path based on framework
     */
    inferComponentPath(spec) {
        const language = this.inferLanguage(spec.framework);
        switch (language) {
            case 'swift':
                return `Sources/${spec.name}/${spec.name}.swift`;
            case 'typescript':
                return `src/${spec.name}.ts`;
            case 'javascript':
                return `src/${spec.name}.js`;
            case 'python':
                return `${spec.name.toLowerCase()}.py`;
            case 'java':
                return `src/main/java/${spec.name}.java`;
            case 'kotlin':
                return `src/main/kotlin/${spec.name}.kt`;
            default:
                return `${spec.name}.txt`;
        }
    }
    /**
     * Infer test file path based on framework
     */
    inferTestPath(spec) {
        const language = this.inferLanguage(spec.framework);
        switch (language) {
            case 'swift':
                return `Tests/${spec.name}Tests/${spec.name}Tests.swift`;
            case 'typescript':
                return `src/__tests__/${spec.name}.test.ts`;
            case 'javascript':
                return `src/__tests__/${spec.name}.test.js`;
            case 'python':
                return `tests/test_${spec.name.toLowerCase()}.py`;
            case 'java':
                return `src/test/java/${spec.name}Test.java`;
            case 'kotlin':
                return `src/test/kotlin/${spec.name}Test.kt`;
            default:
                return `${spec.name}Test.txt`;
        }
    }
    /**
     * Infer programming language from framework
     */
    inferLanguage(framework) {
        const lowerFramework = framework.toLowerCase();
        if (lowerFramework.includes('swift') || lowerFramework.includes('ios')) {
            return 'swift';
        }
        else if (lowerFramework.includes('typescript') || lowerFramework.includes('ts')) {
            return 'typescript';
        }
        else if (lowerFramework.includes('javascript') || lowerFramework.includes('js')) {
            return 'javascript';
        }
        else if (lowerFramework.includes('python')) {
            return 'python';
        }
        else if (lowerFramework.includes('java') && !lowerFramework.includes('javascript')) {
            return 'java';
        }
        else if (lowerFramework.includes('kotlin')) {
            return 'kotlin';
        }
        return 'typescript'; // Default
    }
}
