// @mdp-framework/gateway/dist/aoi/followup-pr/migration-generator.js
/**
 * Migration Generator
 *
 * Generates code migrations to replace custom code with framework components.
 */
export class MigrationGenerator {
    aiProvider;
    constructor(aiProvider) {
        this.aiProvider = aiProvider;
    }
    /**
     * Generate migration code to replace custom implementation with framework component
     */
    async generateMigration(originalPR, spec) {
        const prompt = this.buildMigrationPrompt(originalPR, spec);
        const response = await this.aiProvider.complete({
            prompt,
            temperature: 0.3,
            max_tokens: 3000,
            schema: this.getMigrationSchema(),
        });
        const migration = response.structured;
        // Convert to PRFileChange format
        return migration.changes.map((change) => ({
            path: change.file,
            operation: change.operation,
            content: change.content || '',
        }));
    }
    /**
     * Build migration prompt
     */
    buildMigrationPrompt(originalPR, spec) {
        const violationFile = spec.metadata.trigger_violation.file;
        const violationMessage = spec.metadata.trigger_violation.message;
        return `You are migrating code to use a new framework component.

Original PR:
Title: ${originalPR.title}
Description: ${originalPR.body || 'No description'}

Policy Violation:
File: ${violationFile}
Message: ${violationMessage}

New Framework Component:
Name: ${spec.name}
Description: ${spec.description}

Specification:
${spec.specification}

Task:
Generate code changes to migrate the custom implementation in the original PR to use the new \`${spec.name}\` component from the framework.

Requirements:
1. Replace custom code with framework component usage
2. Update imports to include the new component
3. Maintain the same functionality
4. Add comments explaining the migration
5. Delete any obsolete files

Output JSON with:
- changes: array of {file: string, operation: 'modify' | 'delete', content?: string, reason: string}
- summary: string (brief migration summary)`;
    }
    /**
     * Get migration schema
     */
    getMigrationSchema() {
        return {
            type: 'object',
            properties: {
                changes: {
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            file: { type: 'string' },
                            operation: { type: 'string', enum: ['modify', 'delete'] },
                            content: { type: 'string' },
                            reason: { type: 'string' },
                        },
                        required: ['file', 'operation', 'reason'],
                    },
                },
                summary: { type: 'string' },
            },
            required: ['changes', 'summary'],
        };
    }
}
