// @mdp-framework/gateway/dist/aoi/followup-pr/pr-creator.js
/**
 * Follow-up PR Creator
 *
 * Creates follow-up PRs to migrate original code to use framework components.
 */
import { MigrationGenerator } from './migration-generator.js';
export class FollowupPRCreator {
    aiProvider;
    githubClient;
    migrationGenerator;
    constructor(aiProvider, githubClient) {
        this.aiProvider = aiProvider;
        this.githubClient = githubClient;
        this.migrationGenerator = new MigrationGenerator(aiProvider);
    }
    /**
     * Create follow-up PR to migrate code
     */
    async createFollowupPR(originalPR, counterPR, spec) {
        try {
            // 1. Generate migration changes
            const changes = await this.migrationGenerator.generateMigration(originalPR, spec);
            // 2. Create branch
            const branchName = `aoi-followup-${spec.id}`;
            const branch = await this.githubClient.createBranch(originalPR.base.repo.full_name, branchName, originalPR.base.ref);
            // 3. Commit migration changes
            await this.githubClient.commitFiles(branch, changes, `Migrate to ${spec.name} framework component`, originalPR.base.repo.full_name);
            // 4. Create PR
            const pr = await this.githubClient.createPullRequest({
                repo: originalPR.base.repo.full_name,
                head: branch,
                base: originalPR.base.ref,
                title: `[AOI] Migrate to ${spec.name} component`,
                body: this.buildFollowupPRDescription(originalPR, counterPR, spec),
            });
            return {
                type: 'create_followup_pr',
                status: 'completed',
                target: {
                    platform: 'github',
                    resource_id: pr.number.toString(),
                    url: pr.html_url,
                },
                payload: { spec, original_pr: originalPR, counter_pr: counterPR, followup_pr: pr },
                metadata: {
                    created_at: new Date().toISOString(),
                    completed_at: new Date().toISOString(),
                },
            };
        }
        catch (error) {
            return {
                type: 'create_followup_pr',
                status: 'failed',
                target: {
                    platform: 'github',
                },
                payload: { spec, original_pr: originalPR, counter_pr: counterPR },
                metadata: {
                    created_at: new Date().toISOString(),
                    error: error instanceof Error ? error.message : String(error),
                },
            };
        }
    }
    /**
     * Build follow-up PR description
     */
    buildFollowupPRDescription(originalPR, counterPR, spec) {
        return `## AOI Follow-up: Migrate to ${spec.name}

This PR migrates the custom code from PR #${originalPR.number} to use the new \`${spec.name}\` framework component.

### Background

The original PR (#${originalPR.number}) violated the policy \`${spec.metadata.policy_id}\`:
> ${spec.metadata.trigger_violation.message}

### Solution

A new framework component \`${spec.name}\` has been implemented in ${counterPR.html_url}.

This PR migrates the code to use that component instead of the custom implementation.

### Component Details

**${spec.name}**
${spec.description}

### Changes

This PR includes:
- Migration of custom code to use \`${spec.name}\`
- Updated imports and dependencies
- Removal of obsolete custom implementations

### References

- Original PR: #${originalPR.number}
- Framework Component PR: ${counterPR.html_url}
- Policy: \`${spec.metadata.policy_id}\`

---

This PR was automatically generated by the AOI Coordinator.
Generated at: ${new Date().toISOString()}
`;
    }
}
