// @mdp-framework/mcp-server/dist/tools/evidence-tools.js
/**
 * Evidence Tools - MCP tools for recording evidence in capsules
 */
/**
 * Evidence Record Tool
 *
 * Attach evidence to a state capsule
 */
export class EvidenceRecordTool {
    store;
    name = 'evidence.record';
    description = 'Attach evidence to a state capsule. Evidence can be files, links, artifacts, logs, or custom data. The capsule must already exist.';
    inputSchema = {
        type: 'object',
        properties: {
            capsule_id: {
                type: 'string',
                description: 'Capsule ID to attach evidence to',
            },
            evidence: {
                type: 'object',
                properties: {
                    type: {
                        type: 'string',
                        enum: ['file', 'link', 'artifact', 'log', 'custom'],
                        description: 'Type of evidence',
                    },
                    ref: {
                        type: 'string',
                        description: 'Reference to the evidence (URI, file path, URL, etc.)',
                    },
                    title: {
                        type: 'string',
                        description: 'Human-readable title for the evidence',
                    },
                    metadata: {
                        type: 'object',
                        description: 'Additional metadata for the evidence',
                    },
                },
                required: ['type', 'ref'],
            },
        },
        required: ['capsule_id', 'evidence'],
    };
    constructor(store) {
        this.store = store;
    }
    async handle(params) {
        const startTime = Date.now();
        try {
            // First, check if the capsule exists
            const exists = await this.store.exists(params.capsule_id);
            if (!exists) {
                return {
                    success: false,
                    error: {
                        code: 'NOT_FOUND',
                        message: `Capsule not found: ${params.capsule_id}`,
                    },
                    metadata: {
                        latency_ms: Date.now() - startTime,
                        cached: false,
                    },
                };
            }
            // Create the evidence object
            const evidence = {
                type: params.evidence.type,
                ref: params.evidence.ref,
                title: params.evidence.title,
                metadata: params.evidence.metadata,
            };
            // Merge the evidence into the capsule
            // We use merge to append to the evidence array
            const mergeResult = await this.store.merge(params.capsule_id, {
                evidence: [evidence], // This will be appended to existing evidence array
                updated_at: new Date().toISOString(),
            });
            if (!mergeResult.ok) {
                return {
                    success: false,
                    error: {
                        code: mergeResult.error.code,
                        message: mergeResult.error.message,
                    },
                    metadata: {
                        latency_ms: Date.now() - startTime,
                        cached: false,
                    },
                };
            }
            return {
                success: true,
                data: {
                    capsule_id: params.capsule_id,
                    evidence_added: evidence,
                    capsule: mergeResult.data,
                },
                metadata: {
                    latency_ms: Date.now() - startTime,
                    cached: false,
                },
            };
        }
        catch (error) {
            return {
                success: false,
                error: {
                    code: 'INTERNAL_ERROR',
                    message: error instanceof Error ? error.message : 'Unknown error',
                },
                metadata: {
                    latency_ms: Date.now() - startTime,
                    cached: false,
                },
            };
        }
    }
}
