// @mdp-framework/gateway/dist/aoi/similarity/component-index.js
/**
 * Component Index
 *
 * Indexes framework components with embeddings for similarity search.
 */
export class ComponentIndex {
    aiProvider;
    components = new Map();
    constructor(aiProvider) {
        this.aiProvider = aiProvider;
    }
    /**
     * Index components for a framework
     */
    async indexComponents(framework, components) {
        // Generate embeddings for all component descriptions
        const descriptions = components.map((c) => `${c.name}: ${c.description}`);
        const embeddings = await this.aiProvider.embed(descriptions);
        // Store components with embeddings
        const indexedComponents = components.map((c, i) => ({
            name: c.name,
            description: c.description,
            location: c.location,
            embedding: embeddings[i].vector,
            framework,
        }));
        this.components.set(framework, indexedComponents);
    }
    /**
     * Get components for a framework
     */
    getComponents(framework) {
        return this.components.get(framework) || [];
    }
    /**
     * Check if framework is indexed
     */
    hasFramework(framework) {
        return this.components.has(framework);
    }
    /**
     * Clear index
     */
    clear() {
        this.components.clear();
    }
    /**
     * Get all indexed frameworks
     */
    listFrameworks() {
        return Array.from(this.components.keys());
    }
}
