// @mdp-framework/ai-services/dist/index.js
/**
 * @mdp-framework/ai-services
 *
 * Provider-agnostic AI service layer.
 *
 * Stream 1 implementation in progress.
 * See: Docs/contracts/STREAM-1-AI-SERVICES.md
 */
// Export contracts (FROZEN interfaces)
export * from './contracts/index.js';
// Providers
export { AnthropicProvider } from './providers/anthropic.js';
// OpenAI provider is optional - import directly from './providers/openai.js' if needed
// export { OpenAIProvider } from './providers/openai.js';
// export type { OpenAIProviderOptions } from './providers/openai.js';
export { MockProvider } from './providers/mock.js';
// Registry
export { DefaultAIProviderRegistry } from './registry.js';
// Prompt templates
export { PromptTemplate, PromptTemplateBuilder, CommonTemplates } from './prompts/template.js';
// Cost tracking
export { CostTracker, getGlobalCostTracker, resetGlobalCostTracker } from './cost-tracker.js';
/**
 * IMPLEMENTATION PROGRESS (Stream 1): ✅ COMPLETE
 *
 * Part 1: Anthropic Provider & Registry
 * ✅ src/providers/anthropic.ts - AnthropicProvider implementation
 * ✅ src/registry.ts - DefaultAIProviderRegistry implementation
 * ✅ __tests__/providers/anthropic.test.ts - Comprehensive tests
 * ✅ __tests__/registry.test.ts - Registry tests
 *
 * Part 2: OpenAI & Mock Providers
 * ✅ src/providers/openai.ts - OpenAIProvider implementation
 * ✅ src/providers/mock.ts - MockProvider for deterministic testing
 * ✅ __tests__/providers/openai.test.ts - OpenAI tests
 * ✅ __tests__/providers/mock.test.ts - Mock provider tests
 *
 * Part 3: Embeddings, Templates & Cost Tracking
 * ✅ src/prompts/template.ts - Prompt templating with validation
 * ✅ src/cost-tracker.ts - Token counting & budget enforcement
 * ✅ __tests__/prompts/template.test.ts - Template tests
 * ✅ __tests__/cost-tracker.test.ts - Cost tracking tests
 * ✅ __tests__/integration.test.ts - End-to-end integration tests
 *
 * SUCCESS CRITERIA:
 * ✅ Token cost estimation within 5%
 * ✅ Embeddings support (OpenAI + Mock providers)
 * ✅ Comprehensive test coverage
 * ✅ All providers implement AIProvider interface
 * ✅ Registry supports multiple providers
 * ✅ Cost tracking with budget enforcement
 * ✅ Prompt templating with type validation
 *
 * TODO-STREAM-1: Stream 1 implementation complete, ready for dependent streams
 */
