// @mdp-framework/ai-services/dist/registry.js
/**
 * AI Provider Registry
 *
 * Manages registration and retrieval of AIProvider implementations.
 * Supports multiple providers with default selection.
 */
export class DefaultAIProviderRegistry {
    providers = new Map();
    defaultProvider = null;
    /**
     * Register a provider
     */
    register(provider) {
        this.providers.set(provider.name, provider);
        // Set as default if it's the first provider
        if (this.defaultProvider === null) {
            this.defaultProvider = provider.name;
        }
    }
    /**
     * Get provider by name
     */
    get(name) {
        const provider = this.providers.get(name);
        if (!provider) {
            throw new Error(`Provider "${name}" not found. Available providers: ${this.listProviders().join(', ')}`);
        }
        return provider;
    }
    /**
     * List all registered provider names
     */
    listProviders() {
        return Array.from(this.providers.keys());
    }
    /**
     * Get the default provider
     */
    getDefault() {
        if (this.defaultProvider === null) {
            throw new Error('No providers registered');
        }
        return this.get(this.defaultProvider);
    }
    /**
     * Check if a provider is registered
     */
    has(name) {
        return this.providers.has(name);
    }
    /**
     * Set the default provider
     */
    setDefault(name) {
        if (!this.has(name)) {
            throw new Error(`Cannot set default: provider "${name}" not registered`);
        }
        this.defaultProvider = name;
    }
    /**
     * Unregister a provider
     */
    unregister(name) {
        this.providers.delete(name);
        // Clear default if it was the removed provider
        if (this.defaultProvider === name) {
            this.defaultProvider =
                this.providers.size > 0
                    ? this.providers.keys().next().value
                    : null;
        }
    }
    /**
     * Clear all providers
     */
    clear() {
        this.providers.clear();
        this.defaultProvider = null;
    }
}
