// @mdp-framework/schemas/dist/index.js
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
// Import JSON schemas (Node.js v20.10+ uses 'with' instead of 'assert')
import policyCardSchema from './policy-card.schema.json' with { type: 'json' };
import stateCapsuleSchema from './state-capsule.schema.json' with { type: 'json' };
import envelopeSchema from './envelope.schema.json' with { type: 'json' };
import aoiSpecSchema from './aoi-spec.schema.json' with { type: 'json' };
import aiEvaluationMetadataSchema from './ai-evaluation-metadata.schema.json' with { type: 'json' };
import agentHandoffSchema from './agent-handoff.schema.json' with { type: 'json' };
import workflowResultSchema from './workflow-result.schema.json' with { type: 'json' };
// Create Ajv instance with formats
const ajv = new Ajv({
    allErrors: true,
    verbose: true,
    strict: false,
});
addFormats(ajv);
// Compile validators
const validatePolicyCard = ajv.compile(policyCardSchema);
const validateStateCapsule = ajv.compile(stateCapsuleSchema);
const validateEnvelope = ajv.compile(envelopeSchema);
const validateAOISpec = ajv.compile(aoiSpecSchema);
const validateAIEvaluationMetadata = ajv.compile(aiEvaluationMetadataSchema);
const validateAgentHandoff = ajv.compile(agentHandoffSchema);
const validateWorkflowResult = ajv.compile(workflowResultSchema);
// Validation helper
function formatErrors(validate) {
    if (!validate.errors)
        return [];
    return validate.errors.map((err) => ({
        path: err.instancePath || '/',
        message: err.message || 'Validation failed',
        keyword: err.keyword,
        params: err.params,
    }));
}
// Exported validators
export const validators = {
    policyCard: {
        validate: (data) => {
            const valid = validatePolicyCard(data);
            return valid;
        },
        errors: () => formatErrors(validatePolicyCard),
    },
    stateCapsule: {
        validate: (data) => {
            const valid = validateStateCapsule(data);
            return valid;
        },
        errors: () => formatErrors(validateStateCapsule),
    },
    envelope: {
        validate: (data) => {
            const valid = validateEnvelope(data);
            return valid;
        },
        errors: () => formatErrors(validateEnvelope),
    },
    aoiSpec: {
        validate: (data) => {
            const valid = validateAOISpec(data);
            return valid;
        },
        errors: () => formatErrors(validateAOISpec),
    },
    aiEvaluationMetadata: {
        validate: (data) => {
            const valid = validateAIEvaluationMetadata(data);
            return valid;
        },
        errors: () => formatErrors(validateAIEvaluationMetadata),
    },
    agentHandoff: {
        validate: (data) => {
            const valid = validateAgentHandoff(data);
            return valid;
        },
        errors: () => formatErrors(validateAgentHandoff),
    },
    workflowResult: {
        validate: (data) => {
            const valid = validateWorkflowResult(data);
            return valid;
        },
        errors: () => formatErrors(validateWorkflowResult),
    },
};
// Export schemas
export const schemas = {
    policyCard: policyCardSchema,
    stateCapsule: stateCapsuleSchema,
    envelope: envelopeSchema,
    aoiSpec: aoiSpecSchema,
    aiEvaluationMetadata: aiEvaluationMetadataSchema,
    agentHandoff: agentHandoffSchema,
    workflowResult: workflowResultSchema,
};
