// @mdp-framework/gateway/dist/evaluators/confidence-scorer.js
/**
 * Confidence Scorer (Stream 2)
 *
 * Utilities for calculating and filtering violations based on AI confidence scores.
 * Helps reduce false positives by downgrading or filtering low-confidence violations.
 */
/**
 * Confidence thresholds for different actions
 */
export const CONFIDENCE_THRESHOLDS = {
    /**
     * Violations below this are filtered out entirely
     */
    MIN_REPORTABLE: 0.5,
    /**
     * Violations below this are downgraded to warnings
     */
    RELIABLE: 0.7,
    /**
     * Violations above this are considered high confidence
     */
    HIGH: 0.9,
};
/**
 * Confidence scoring utilities
 */
export class ConfidenceScorer {
    /**
     * Calculate average confidence across violations
     */
    static calculateAverageConfidence(violations) {
        if (violations.length === 0) {
            return 1.0; // No violations = perfect confidence
        }
        const sum = violations.reduce((acc, v) => acc + (v.confidence ?? 1.0), 0);
        return sum / violations.length;
    }
    /**
     * Calculate weighted confidence score
     * Errors are weighted more heavily than warnings
     */
    static calculateWeightedConfidence(violations) {
        if (violations.length === 0) {
            return 1.0;
        }
        const weights = { error: 1.0, warning: 0.5, info: 0.25 };
        let totalWeight = 0;
        let weightedSum = 0;
        for (const violation of violations) {
            const confidence = violation.confidence ?? 1.0;
            const weight = weights[violation.severity];
            weightedSum += confidence * weight;
            totalWeight += weight;
        }
        return totalWeight > 0 ? weightedSum / totalWeight : 1.0;
    }
    /**
     * Filter violations by confidence threshold
     *
     * - Violations below MIN_REPORTABLE are removed
     * - Violations below RELIABLE are downgraded to warnings (if they were errors)
     */
    static filterByConfidence(violations, threshold = CONFIDENCE_THRESHOLDS.RELIABLE) {
        return violations
            .filter(v => {
            // Remove violations below minimum threshold
            const confidence = v.confidence ?? 1.0;
            return confidence >= CONFIDENCE_THRESHOLDS.MIN_REPORTABLE;
        })
            .map(v => {
            const confidence = v.confidence ?? 1.0;
            // Downgrade low-confidence errors to warnings
            if (confidence < threshold && v.severity === 'error') {
                return {
                    ...v,
                    severity: 'warning',
                    message: `${v.message} (confidence: ${(confidence * 100).toFixed(0)}%)`,
                    metadata: {
                        ...v.metadata,
                        original_severity: 'error',
                        downgraded_reason: 'low_confidence',
                    },
                };
            }
            return v;
        });
    }
    /**
     * Categorize violations by confidence level
     */
    static categorizeByConfidence(violations) {
        const high = [];
        const medium = [];
        const low = [];
        for (const violation of violations) {
            const confidence = violation.confidence ?? 1.0;
            if (confidence >= CONFIDENCE_THRESHOLDS.HIGH) {
                high.push(violation);
            }
            else if (confidence >= CONFIDENCE_THRESHOLDS.RELIABLE) {
                medium.push(violation);
            }
            else if (confidence >= CONFIDENCE_THRESHOLDS.MIN_REPORTABLE) {
                low.push(violation);
            }
        }
        return { high, medium, low };
    }
    /**
     * Get confidence summary for a result
     */
    static getConfidenceSummary(result) {
        const categorized = this.categorizeByConfidence(result.violations);
        return {
            average: this.calculateAverageConfidence(result.violations),
            weighted: this.calculateWeightedConfidence(result.violations),
            distribution: {
                high: categorized.high.length,
                medium: categorized.medium.length,
                low: categorized.low.length,
            },
        };
    }
    /**
     * Check if result should trigger action based on confidence
     *
     * Returns true if:
     * - Any error violations with confidence >= RELIABLE
     * - Multiple warning violations with confidence >= RELIABLE
     */
    static shouldTriggerAction(result) {
        const reliableViolations = result.violations.filter(v => (v.confidence ?? 1.0) >= CONFIDENCE_THRESHOLDS.RELIABLE);
        // Any reliable error should trigger
        if (reliableViolations.some(v => v.severity === 'error')) {
            return true;
        }
        // Multiple reliable warnings should trigger
        const reliableWarnings = reliableViolations.filter(v => v.severity === 'warning');
        return reliableWarnings.length >= 2;
    }
}
